source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -0,0 +1,260 @@
+
+    // Check for duplicates.
+    if (sk_X509_NAME_find(list.get(), nullptr, name)) {
+      continue;
+    }
+
+    bssl::UniquePtr<X509_NAME> name_dup(X509_NAME_dup(name));
+    if (name_dup == nullptr || !sk_X509_NAME_push(list.get(), name_dup.release())) {
+      throw EnvoyException(absl::StrCat("Failed to load trusted client CA certificate"));
+    }
+  }
+  SSL_CTX_set_client_CA_list(ctx, list.release());
+}
+
+void SPIFFEValidator::updateDigestForSessionId(bssl::ScopedEVP_MD_CTX& md,
+                                               uint8_t hash_buffer[EVP_MAX_MD_SIZE],
+                                               unsigned hash_length) {
+  int rc;
+  for (auto& ca : ca_certs_) {
+    rc = X509_digest(ca.get(), EVP_sha256(), hash_buffer, &hash_length);
+    RELEASE_ASSERT(rc == 1, Utility::getLastCryptoError().value_or(""));
+    RELEASE_ASSERT(hash_length == SHA256_DIGEST_LENGTH,
+                   fmt::format("invalid SHA256 hash length {}", hash_length));
+    rc = EVP_DigestUpdate(md.get(), hash_buffer, hash_length);
+    RELEASE_ASSERT(rc == 1, Utility::getLastCryptoError().value_or(""));
+  }
+}
+
+int SPIFFEValidator::initializeSslContexts(std::vector<SSL_CTX*>, bool) {
+  return SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
+}
+
+int SPIFFEValidator::doVerifyCertChain(X509_STORE_CTX* store_ctx,
+                                       Ssl::SslExtendedSocketInfo* ssl_extended_info,
+                                       X509& leaf_cert, const Network::TransportSocketOptions*) {
+  if (!SPIFFEValidator::certificatePrecheck(&leaf_cert)) {
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(Envoy::Ssl::ClientValidationStatus::Failed);
+    }
+    stats_.fail_verify_error_.inc();
+    return 0;
+  }
+
+  auto trust_bundle = getTrustBundleStore(&leaf_cert);
+  if (!trust_bundle) {
+    if (ssl_extended_info) {
+      ssl_extended_info->setCertificateValidationStatus(Envoy::Ssl::ClientValidationStatus::Failed);
+    }
+    stats_.fail_verify_error_.inc();
+    return 0;
+  }
+
+  // Set the trust bundle's certificate store on the context, and do the verification.
+  store_ctx->ctx = trust_bundle;
+  auto ret = X509_verify_cert(store_ctx);
+  if (ssl_extended_info) {
+    ssl_extended_info->setCertificateValidationStatus(
+        ret == 1 ? Envoy::Ssl::ClientValidationStatus::Validated
+                 : Envoy::Ssl::ClientValidationStatus::Failed);
+  }
+  if (!ret) {
+    stats_.fail_verify_error_.inc();
+  }
+  return ret;
+}
+
+X509_STORE* SPIFFEValidator::getTrustBundleStore(X509* leaf_cert) {
+  bssl::UniquePtr<GENERAL_NAMES> san_names(static_cast<GENERAL_NAMES*>(
+      X509_get_ext_d2i(leaf_cert, NID_subject_alt_name, nullptr, nullptr)));
+  if (!san_names) {
+    return nullptr;
+  }
+
+  std::string trust_domain;
+  for (const GENERAL_NAME* general_name : san_names.get()) {
+    const std::string san = Utility::generalNameAsString(general_name);
```

Comment:
I think this is only checking the first of any SAN type, but from the spec:
>  An X.509 SVID MUST contain exactly one URI SAN. It MAY contain any number of other SAN fields, including DNS SANs.

I think we want something like the following:
```suggestion
  for (const GENERAL_NAME* general_name : san_names.get()) {
     if (generate_name->type != GEN_URI) {
         continue;
     }
    const std::string san = Utility::generalNameAsString(general_name);
```

Reply:
Thanks! resolved in https://github.com/envoyproxy/envoy/pull/14884/commits/e6b2c0fdca1e411e814d097457f958b8dda5b84c